361 dist quantiles null

dajmcdon opened the pull request:
### Checklist

Please:

- [x] Make sure this PR is against "dev", not "main".
- [x] Request a review from one of the current epipredict main reviewers
- [x] Make sure to bump the version number in `DESCRIPTION` and `NEWS.md`.
      Always increment the patch version number (the third number), unless you are
      making a release PR from dev to main, in which case increment the minor
      version number (the second number).
- [x] Describe changes made in NEWS.md, making sure breaking changes
      (backwards-incompatible changes to the documented interface) are noted.
      Collect the changes under the next release number (e.g. if you are on
      0.7.2, then write your changes under the 0.8 heading).

### Magic GitHub syntax to mark associated Issue(s) as resolved when this is merged into the default branch

- Resolves #361

brookslogan commented on R/dist_quantiles.R:
```diff
@@ -5,6 +5,9 @@ new_quantiles <- function(values = double(), quantile_levels = double()) {
   vec_cast(values, double())
   vec_cast(quantile_levels, double())
```
(this is a no-op. `vec_cast` doesn't have side effects)

dajmcdon replied:
I'm not positive this is the "best" thing to do. It doesn't need to convert. I'm open to suggestions.

brookslogan replied:
I'm not sure.  It certainly seems simpler to think about if we could go ahead and cast.  But I'm not sure if there's a situation where someone really really wants their integer-type/esoteric-yet-still-double-compatible quantiles.

brookslogan commented on R/dist_quantiles.R:
```diff
@@ -5,6 +5,9 @@ new_quantiles <- function(values = double(), quantile_levels = double()) {
   stopifnot(length(values) == length(quantile_levels))
+  if (length(values) == 0L) {
+    return(distributional::dist_degenerate(double()))
+  }
```
question: is this out of order vs. line 7?

question: why not a 0-length `dist_quantiles`?

dajmcdon replied:
Right. I don't think we should ever hit this line. But I put it in anyway.

brookslogan commented on R/dist_quantiles.R:
```diff
@@ -5,6 +5,9 @@ new_quantiles <- function(values = double(), quantile_levels = double()) {
   vec_cast(quantile_levels, double())
   stopifnot(length(values) == length(quantile_levels))
+  if (length(values) == 0L) {
+    return(distributional::dist_degenerate(double()))
```
issue: this is the wrong type to come out of `new_quantiles()`; `dist_degenerate(double())` is a `"distribution"`, not an element thereof.  Plus it seems like asking for trouble anyway to have these ill-formed distributions around.

```suggestion
    cli_abort("new_quantiles() requires at least one quantile to be specified.  If you meant to make an empty list of distributions, try dist_quantiles() instead.")
```

dajmcdon replied:
agreed. Also changing the function defaults to `double(1)` to avoid triggering by default.

brookslogan commented on R/dist_quantiles.R:
```diff
@@ -60,6 +63,9 @@ dist_quantiles <- function(values, quantile_levels) {
+  if (length(args$values) == 0L) {
+    return(distributional::dist_degenerate(double()))
+  }
   qntls <- as_list_of(map2(args$values, args$quantile_levels, new_quantiles))
```
```suggestion
  qntls <- as_list_of(map2(args$values, args$quantile_levels, new_quantiles),
                      .ptype = new_quantiles(NA_real_, 0.5))
```

brookslogan replied:
or
```suggestion
  qntls <- map2(args$values, args$quantile_levels, new_quantiles)
```

dajmcdon replied:
Going with the first version.

brookslogan commented on R/dist_quantiles.R:
```diff
@@ -5,6 +5,9 @@ new_quantiles <- function(values = double(), quantile_levels = double()) {
   vec_cast(values, double())
   vec_cast(quantile_levels, double())
   stopifnot(length(values) == length(quantile_levels))
```
```suggestion
  stopifnot(length(values) == length(quantile_levels))
  values <- unname(values)
```
this can prevent `vctrs::vec_cast_common` from producing things like
```
Error:
! Can't convert `..1` <dist_quantiles> to <dist_quantiles>
```
but you'll need to accept/review the snapshot tests.

dajmcdon replied:
Will do.

brookslogan reviewed (approved):
- I think `new_quantiles` needs a small fixup to be a bit more consistent/demanding.
- I've suggested an alternative approach that's mostly stylistic, but also involves avoiding some unclear vctrs messages that we might encounter in further development.

dajmcdon commented:
@brookslogan I think I've addressed all your suggestions, but maybe take a quick look if you can. I also added some code to compute weighted interval score for these distributions (which we need to use in various places). Sorry!

brookslogan reviewed (approved):
Left a few more thoughts.  Like the last batch, all sort of preference-based.

dajmcdon commented on R/weighted_interval_score.R:
```diff
@@ -0,0 +1,73 @@
+
+#' @export
+weighted_interval_score.default <- function(x, actual, ...) {
+  cli_abort("Weighted interval score can only be calculated if `x` has class {.cls distribution}.")
```
A `distribution` is a vector container that can contain 1 or more `dist_quantiles`. But it can contain different types of distributions simultaneously. Like:
``` r
library(distributional)
dstn <- c(dist_normal(), dist_quantiles(1:4, 1:4/5))
```
